### ISSUE TITLE
Define a proper deprecation policy

### ISSUE THREAD

**Issue by ambs** [PerlDancer/Dancer2#743]
Hello

There was some "discussion" between me and @xsawyerx about deprecation of `param`. Not that we disagree on deprecating it, but rather, we disagree on what is deprecation.

So, this thread is to ask for a proper discussion on what is deprecation for Dancer2.

In my point of view, the definition of deprecation we had in the old days for Dancer was working well:

   1 First, the deprecated functionality/feature was marked as **soft deprecated**. It still worked, but issued a warning to the log files.
   2 Then, after a period of time, or after a fixed number of releases, the functionality/feature was marked as **hard deprecated**. It no longer worked, but an error was thrown on its usage.
   3 Again, after a period of time, or after a fixed number of releases, the functionality/feature disappeared completely.

This allowed us to clean code that we think is bad, buggy or that makes confusion to unaware developers. Also, it would allow us to, for example when deprecating a DSL keyword, to have it free again after some time, in order to fix it properly, for example.

In that "discussion" I had with @xsawyerx , his idea was to leave the `param` DSL keyword hanging in the code, but issuing a warning. As I've told him, for apps that are running, developers will not notice a feature is deprecated unless it gets to the **hard deprecated** stage, where it fails. In other words, it is very not probable they will look into the app logs unless it is failing. If we keep the code warning for ever, what we are doing is making our log files bigger (a lot!).

Therefore, I ask that we think about deprecation, and define a proper deprecation policy that should be followed by any feature, keyword or submodule that we want to change or remove.

**Comment by VinceW** [comment]
Closely related to deprecation is in my opinion backward-compatibility.

The perl.org documentation is pretty clear about this at: http://perldoc.perl.org/perlpolicy.html#BACKWARD-COMPATIBILITY-AND-DEPRECATION

I think a Dancer2 policy should adhere to this policy because companies, project managers, change managers, developers, etc who are familiar with Perl (but not with Dancer2) are probably expecting the same compatibility/deprecation policy as Perl in general has.

In respond to the scenario outlined by @ambs:
> As I've told him, for apps that are running, developers will not notice a feature is deprecated unless it gets to the hard deprecated stage, where it fails. In other words, it is very not probable they will look into the app logs unless it is failing. If we keep the code warning for ever, what we are doing is making our log files bigger (a lot!).

*Normally* in a development cycle, running apps are not changing. If, however, there is a need to apply changes to the app this isn't normally **not** done in a production environment. Therefore there is at least a development environment where the changes are developed. In a development environment, the developer *should* test the changes. At minimum the developer should check the logs on errors and/or warnings.

In my opinion: All changes to apps should be done in a develop(ers)(ment) environment and be tested before put into production.

Following this, Dancer2 should provide means to inform/warn/error a developer and/or a tester. Seeing the above from @ambs I think this is (more or less) in place, but never put in a formal perspective/document/boundary.

Another aspect what could be mentioned in this kind of policy is the support. How long is a version supported by the Dancer2 team? This kind of information could be relevant in the decision process when a version is to be picked.

until here my 5 cents, looking forward to an interesting discussion about this.

Best,
VinceW

**Comment by rleir** [comment]
@VinceW sums this up nicely. In addition, I think that any time we deprecate something there should be an issue raised here for discussion. Then we can think of things like
- how severe would breakage be due to continued use of the hard deprecated feature? (data corrupted, site down, visual difference, minor bother)
- how many existing projects are likely to be affected
- how much work might be needed in a project to replace the use of the deprecated whatever
- benefit: how much are we gaining by migrating away from the deprecated feature

There is always some contention between devs proposing change (this new thing is essential!) and devs using the existing version (no way! do not change that). One solution is to fork the project, I.E. create Dancer2 from Dancer!

@ambs my preference is that Hard Deprecated should not intentionally disable the feature. It still exists, though it is expected that there may be increasing breakage when it is used.

This issue is stale, can we close it? When you get a moment, thanks.

**Comment by cromedome** [comment]
Bumping this issue with a comment to raise some awareness.

As I go through the backlog of old issues, look through code, and add documentation, I see a lot of places where we intended to deprecate functionality (or, *should* deprecate something) but haven't for lack of policy.

D2 has been around long enough, accumulated some cruft, and has been stable long enough we should start trimming out the things we got wrong at the beginning. I like what @ambs has proposed, and would like to codify it further:

At the time I write this, the current version of Dancer2 is 0.300006. Anything we mark for soft deprecation now would not be hard deprecated until two major revisions for now (so, 0.500000). At the rate at which we release D2, this would be *plenty* of time for people to migrate to whatever feature/fix is replacing the deprecated functionality. 

I am all for hearing the community out regarding deprecated features per @rleir's suggestion. I think that is super important, I'm sure there are always going to be use cases the core team hasn't considered, but I think the decision to deprecate should ultimately continue to lie with the core team.

Dancers, what say you?
